src/metabase/lib/fe_util.cljc

```diff
@@ -71,54 +87,177 @@
+;; is "Unknown Field" instead of it's id.
+(defn- set-display-name-for-unknown-field
+  [metadata]
+  (if (and (nil? (:id metadata)) (nil? (:lib/expression-name metadata)))
+    (assoc metadata :display-name (i18n/tru "Unknown Field"))
+    metadata))
 
 (defn- column-metadata-from-ref
   [query stage-number a-ref]
-  (lib.filter/add-column-operators
-   (lib.field/extend-column-metadata-from-ref
-    query stage-number
-    (lib.metadata.calculation/metadata query stage-number a-ref)
-    a-ref)))
-
-(mu/defn expression-parts :- ExpressionParts
-  "Return the parts of the filter clause `expression-clause` in query `query` at stage `stage-number`."
-  ([query expression-clause]
-   (expression-parts query -1 expression-clause))
+  (as->
+   (lib.metadata.calculation/metadata query stage-number a-ref) metadata
+    (lib.field/extend-column-metadata-from-ref query stage-number metadata a-ref)
+    (lib.filter/add-column-operators metadata)
+    (set-display-name-for-unknown-field metadata)))
+
+(defn- segment-metadata-from-ref
+  [query segment-ref]
+  (if-let [segment (lib.metadata/segment query (last segment-ref))]
+    segment
+    {:lib/type :metadata/segment
+     :id (last segment-ref)
+     :display-name (i18n/tru "Unknown Segment")}))
+
+(defn- metric-metadata-from-ref
+  [query metric-ref]
+  (if-let [metric (lib.metadata/metric query (last metric-ref))]
+    metric
+    {:lib/type :metadata/metric
+     :id (last metric-ref)
+     :display-name (i18n/tru "Unknown Metric")}))
+
+(defmulti expression-parts-method
+  "Builds the expression parts by dispatching on the type of the argument."
+  {:arglists '([query stage-number arg])}
+  (fn [_query _stage-number value]
+    (lib.dispatch/dispatch-value value))
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod expression-parts-method :default
+  [query stage-number value]
+  (let [[op options & args] (cond-> value
+                              (expandable-temporal-expression? value) expand-temporal-expression
+                              (case-or-if-expression? value) flatten-case-or-if-expression)]
+    {:lib/type :mbql/expression-parts
+     :operator op
+     :options  options
+     :args     (mapv #(expression-parts-method query stage-number %) args)}))
+
+(defmethod expression-parts-method :dispatch-type/expression-parts
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :dispatch-type/string
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :dispatch-type/integer
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :dispatch-type/number
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :dispatch-type/boolean
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :dispatch-type/keyword
+  [_query _stage-number value]
+  (name value))
+
+(defmethod expression-parts-method :dispatch-type/nil
+  [_query _stage-number value]
+  value)
```

Comment:
All of these "identity" cases can be handled in one place with a `doseq`; we can pair on this.